File: python/deals/deals_api_solution.py
import boto3
import json
import base64

deals = [
    {
        'url': 'https://example.com/deals/001',
        'name': 'Deal 001',
        'description': 'Deal 001 description',
        'price': '100',
        'currency': 'USD',
    },
    {
        'url': 'https://example.com/deals/002',
        'name': 'Deal 002',
        'description': 'Deal 002 description',
        'price': '200',
        'currency': 'USD'
    },
    {
        'url': 'https://example.com/deals/003',
        'name': 'Deal 003',
        'description': 'Deal 003 description',
        'price': '300',
        'currency': 'USD'
    }
]

# Function to convert json to xml
def json_to_xml(json_data):
    xml = '<deals>'
    for deal in json_data:
        xml += '<deal>'
        xml += '<url>' + deal['url'] + '</url>'
        xml += '<name>' + deal['name'] + '</name>'
        xml += '<description>' + deal['description'] + '</description>'
        xml += '<price>' + deal['price'] + '</price>'
        xml += '<currency>' + deal['currency'] + '</currency>'
        xml += '</deal>'
    xml += '</deals>'
    return xml

# Lambda function to insert user into DynamoDB table
def lambda_handler(event, context):
    print(event)
    try:
        #read the Accept header to determine the format of the response
        accept = event['headers']['Accept']

        #if the Accept header is application/xml, convert the deals to xml and return it
        if accept == 'application/xml':
            return {
                'statusCode': 200,
                'body': json_to_xml(deals),
                'headers': {
                    'content-type': 'application/xml'
                }
            }
        
        #if the Accept header is application/json, return the deals as json
        elif accept == 'application/json':
            return {
                'statusCode': 200,
                'body': json.dumps(deals),
                'headers': {
                    'content-type': 'application/json'
                }
            }
    except Exception as e:
        print(e)
        return {'statusCode': 500}

File: csharp/src/Deals/DealsApi_Solution.cs
using System.Text;
using System.Text.Json;
using System.Xml.Serialization;
using static Deals.DealsApi;

namespace Deals
{
    public class DealsApi_Solution
    {
        // List of Deals
        List<Deal> deals = new List<Deal>() {
            new Deal {
                Url = "https://example.com/deals/001",
                Name= "Deal 001",
                Description="Deal 001 description",
                Price= "100",
                Currency= "USD"},
            new Deal {
                Url = "https://example.com/deals/002",
                Name= "Deal 002",
                Description="Deal 002 description",
                Price= "200",
                Currency= "USD"},
            new Deal {
                Url = "https://example.com/deals/003",
                Name= "Deal 003",
                Description="Deal 003 description",
                Price= "300",
                Currency= "USD"}
        };

        //<summary>Method to get deals object</summary>
        //<returns>Returns APIResponse object</returns>  
        public APIResponse GetDeals(string format = "json")
        {
            APIResponse response = new APIResponse();
            if (format == "json")
            {
                response = GetDealsJSON();
            }
            else
            {                
                response = GetDealsXML();
            }
            return response;
        }

        //<summary>Method to serialize deals object to JSON.</summary>       
        //<returns>Returns JSON string</returns>      
        private APIResponse GetDealsJSON()
        {
            var json = JsonSerializer.Serialize(deals);
            return new APIResponse
            {
                ContentType = "application/json",
                Content = json
            };
        }

        #region
        // Create a method to convert Deals object to XML
        //<summary>Method to serialize deals object to XML.</summary>
        //<returns>Returns XML string</returns>
        private APIResponse GetDealsXML()
        {
            var serializer = new XmlSerializer(typeof(List<Deal>));
            var stringWriter = new StringWriter();
            serializer.Serialize(stringWriter, deals);
            return new APIResponse
            {
                ContentType = "application/xml",
                Content = stringWriter.ToString()
            };
        }

        public class Deal
        {
            public string Url { get; set; } = string.Empty;
            public string Name { get; set; } = string.Empty;
            public string Description { get; set; } = string.Empty;
            public string Price { get; set; } = string.Empty;
            public string Currency { get; set; } = string.Empty;
        }
    }
}
